sound/soc/sof/intel/hda-dsp.c

```
@@ -1044,3 +1045,64 @@ int hda_dsp_core_get(struct snd_sof_dev *sdev, int core)
 
 	return ret;
 }
+
+int hda_dsp_ipc4_core_get(struct snd_sof_dev *sdev, int core)
+{
+	struct sof_ipc4_dx_state_info dx_state;
+	struct sof_ipc4_msg msg;
+	int ret;
+
+	ret = hda_dsp_enable_core(sdev, BIT(core));
+	if (ret < 0) {
+		dev_err(sdev->dev, "%s: failed to power up core %d with err: %d\n",
+			__func__, core, ret);
+		return ret;
+	}
+
+	/* No need to send IPC for primary core or if FW boot is not complete */
+	if (sdev->fw_state != SOF_FW_BOOT_COMPLETE || core == SOF_DSP_PRIMARY_CORE)
+		return 0;
+
+	dx_state.core_mask = BIT(core);
+	dx_state.dx_mask = BIT(core);
+	msg.primary = SOF_IPC4_MSG_TYPE_SET(SOF_IPC4_MOD_SET_DX);
+	msg.primary |= SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST);
+	msg.primary |= SOF_IPC4_MSG_TARGET(SOF_IPC4_MODULE_MSG);
+	msg.data_ptr = &dx_state;
+	msg.data_size = sizeof(dx_state);
+
+	ret = sof_ipc_tx_message(sdev->ipc, &msg, msg.data_size, NULL, 0);
+	if (ret) {
+		int err = hda_dsp_core_reset_power_down(sdev, BIT(core));
+
+		if (err < 0)
+			dev_err(sdev->dev,
+				"%s: failed to power down core: %d with err: %d\n",
+				__func__, core, err);
+	}
+
+	return ret;
+}
+
+int hda_dsp_ipc4_core_put(struct snd_sof_dev *sdev, int core)
+{
+	struct sof_ipc4_dx_state_info dx_state;
+	struct sof_ipc4_msg msg;
+	int ret;
+
+	dx_state.core_mask = BIT(core);
+	dx_state.dx_mask = 0;
+	msg.primary = SOF_IPC4_MSG_TYPE_SET(SOF_IPC4_MOD_SET_DX);
+	msg.primary |= SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST);
+	msg.primary |= SOF_IPC4_MSG_TARGET(SOF_IPC4_MODULE_MSG);
+	msg.data_ptr = &dx_state;
+	msg.data_size = sizeof(dx_state);
+
+	ret = sof_ipc_tx_message(sdev->ipc, &msg, msg.data_size, NULL, 0);
+
```

if ret < 0, do we need to return it without doing the following operations ?

---

We ignore the error for the primary core and proceed to turn it off. For the non primary cores we return with the error code, but don't do anything on the host side.
We need to send the SET_DX message for all cores, but the host only manages the primary one.